File: BSvZP-Common/Messages/ConfigurationReply.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using Common;

namespace Messages
{
    public class ConfigurationReply : Reply
    {
        #region Private Properties
        private static Int16 ClassId { get { return (Int16)MESSAGE_CLASS_IDS.ConfigurationReply; } }
        #endregion

        #region Public Properties
        public override Message.MESSAGE_CLASS_IDS MessageTypeId() { return (Message.MESSAGE_CLASS_IDS)ClassId; }

        public GameConfiguration Configuration { get; set; }
        public static new int MinimumEncodingLength
        {
            get
            {
                return 4                // Object header
                       + GameConfiguration.MinimumEncodingLength;
            }
        }
        #endregion
        
        #region Constructors and Factory Methods
        /// <summary>
        /// Default message constructor, used by Factory methods (i.e., the Create methods)
        /// </summary>

        protected ConfigurationReply() { }

        /// <summary>
        /// Constructor used by all specializations, which in turn are used by
        /// senders of a message 
        /// </summary>
        /// <param name="conversationId">conversation id</param>
        /// <param name="status">Status of the ack/nak</status>
        /// <param name="note">error message or note</note>
        public ConfigurationReply(PossibleStatus status, GameConfiguration config, string note = null) :
            base(Reply.PossibleTypes.ConfigurationReply, status, note)
        {
            Configuration = config;
        }

        /// <summary>
        /// Factor method to create a message from a byte list
        /// </summary>
        /// <param name="messageBytes">A byte list from which the message will be decoded</param>
        /// <returns>A new message of the right specialization</returns>
        new public static ConfigurationReply Create(ByteList messageBytes)
        {
            ConfigurationReply result = null;

            if (messageBytes==null || messageBytes.RemainingToRead<MinimumEncodingLength)
                throw new ApplicationException("Invalid message byte array");
            if (messageBytes.PeekInt16() != ClassId)
                throw new ApplicationException("Invalid message class id");
            else
            {
                result = new ConfigurationReply();
                result.Decode(messageBytes);
            }

            return result;
        }
        #endregion

        #region Encoding and Decoding methods

        override public void Encode(ByteList bytes)
        {
            bytes.Add(ClassId);                           // Write out this class id first

            Int16 lengthPos = bytes.CurrentWritePosition;   // Get the current write position, so we
                                                                // can write the length here later
            bytes.Add((Int16) 0);                           // Write out a place holder for the length

            base.Encode(bytes);                             // Encode stuff from base class

            bytes.Add(Configuration);

            Int16 length = Convert.ToInt16(bytes.CurrentWritePosition - lengthPos - 2);
            bytes.WriteInt16To(lengthPos, length);          // Write out the length of this object        

        }

        override public void Decode(ByteList bytes)
        {
            Int16 objType = bytes.GetInt16();
            Int16 objLength = bytes.GetInt16();

            bytes.SetNewReadLimit(objLength);

            base.Decode(bytes);

            Configuration = bytes.GetDistributableObject() as GameConfiguration;
            
            bytes.RestorePreviosReadLimit();
        }

        #endregion

    }
}

File: BSvsZP-Common -Java/Messages/ConfigurationReply.java
package Messages;

import java.io.NotActiveException;
import java.net.UnknownHostException;
import org.omg.CORBA.portable.ApplicationException;
import Common.ByteList;
import Common.GameConfiguration;

public class ConfigurationReply extends Reply {

    public GameConfiguration Configuration;
    private static int MinimumEncodingLength;

    protected ConfigurationReply() {
    }

    public ConfigurationReply(PossibleStatus status, GameConfiguration config, String... note) {
        super(Reply.PossibleTypes.ConfigurationReply, status, ((note.length == 1) ? note[0] : null));
        Configuration = config;
    }

   
    public static ConfigurationReply Create(ByteList messageBytes) throws Exception {
        ConfigurationReply result = null;

        if (messageBytes == null || messageBytes.getRemainingToRead() < getMinimumEncodingLength()) {
            throw new ApplicationException("Invalid message byte array", null);
        }
        if (messageBytes.PeekInt16() !=(short) MESSAGE_CLASS_IDS.ConfigurationReply.getValue()) {
            throw new ApplicationException("Invalid message class id", null);
        } else {
            result = new ConfigurationReply();
            result.Decode(messageBytes);
        }

        return result;
    }

    @Override
    public void Encode(ByteList bytes) throws NotActiveException, UnknownHostException, Exception {
        bytes.Add((short) MESSAGE_CLASS_IDS.ConfigurationReply.getValue());                           // Write out this class id first
        bytes.update();
        short lengthPos = bytes.getCurrentWritePosition();   // Get the current write position, so we
        bytes.update();                                              // can write the length here later
        bytes.Add((short) 0);                           // Write out a place holder for the length
        bytes.update();
        super.Encode(bytes);                             // Encode stuff from base class

        bytes.Add(Configuration);
        bytes.update();
        short length = (short) (bytes.getCurrentWritePosition() - lengthPos - 2);
        bytes.WriteInt16To(lengthPos, length);          // Write out the length of this object        
        bytes.update();
    }

    @Override
    public void Decode(ByteList bytes) throws Exception {
        short objType = bytes.GetInt16();
        bytes.update();
        short objLength = bytes.GetInt16();
        bytes.update();

        bytes.SetNewReadLimit(objLength);
        bytes.update();

        super.Decode(bytes);

        Configuration = (GameConfiguration) bytes.GetDistributableObject();
        bytes.update();

        bytes.RestorePreviosReadLimit();
        bytes.update();
    }

    public GameConfiguration getConfiguration() {
        return Configuration;
    }

    public void setConfiguration(GameConfiguration configuration) {
        Configuration = configuration;
    }

    public static int getMinimumEncodingLength() {
        MinimumEncodingLength = 4 		// Object header
        						+ GameConfiguration.getMinimumEncodingLength();
        return MinimumEncodingLength;
    }

    @Override
    public short getClassId() {
        return (short) MESSAGE_CLASS_IDS.ConfigurationReply.getValue();
    }

    @Override
    public Message.MESSAGE_CLASS_IDS MessageTypeId() {
        return Message.MESSAGE_CLASS_IDS.ConfigurationReply;
    }

    @Override
    public int compareTo(Object arg0) {

        return 0;
    }
}
